File: python/algs/selection_sort.py
"""
Сортировка выбором -- алгоритм сортировки последовательности чисел.
Описание:
1. Проходим в цикле по массив, индекс i до конца массива
2. Фиксируем текущий элемент как минимальный
3. Проходим во вложенном цикле по массиву, j индекс от i до конца массива
4. Сравниваем каждый элемент элемент j с фиксированным элементов i,
заменяем минмиальный элемент
Сложность алгоритма O(n*n)
"""


def selection_sort(arr):
    count = 0  # счетчик для оценки сложности
    for i in range(0, len(arr)):
        min_element_index = i
        for j in range(i, len(arr)):
            if arr[j] < arr[min_element_index]:
                min_element_index = j
        tmp = arr[i]
        arr[i] = arr[min_element_index]
        arr[min_element_index] = tmp
        count += 1
    return arr, count


if __name__ == "__main__":
    arr = [4, 2, 1, 4, 10, 6, 1, 8, 9]
    number = 7
    sorted_arr, count = selection_sort(arr)
    print(sorted_arr)
    print(count)

File: cpp/algs/selection_sort.cpp

#include <cmath>
#include <iostream>
#include <vector>

void selection_sort(std::vector<int>& array)
{
  for (int i = 0; i < array.size(); ++i) {
    for (int j = i + 1; j < array.size(); ++j) {
      if (array[j] < array[i]) {
        int tmp = array[i];
        array[i] = array[j];
        array[j] = tmp;
      }
    }
  }
}

int main()
{
  std::vector<int> array = {0, 3, 2, 1, 4, 5};

  selection_sort(array);
  for (const auto& item : array) {
    std::cout << item << std::endl;
  }

  return 0;
}
